File: examples/playground/android/app/src/main/java/com/yolk/runtime/YolkBin.kt
package com.yolk.runtime

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

/**
 * A custom binary protocol (TLV) that matches the TypeScript and Swift `YolkBin`.
 * This is the exclusive way data is transferred in Yolk's buffer-only architecture.
 */
object YolkBin {
    fun encode(value: Any?): ByteBuffer {
        val size = sizeOf(value)
        val buffer = ByteBuffer.allocateDirect(size).order(ByteOrder.LITTLE_ENDIAN)
        write(value, buffer)
        buffer.flip()
        return buffer
    }

    fun decode(buffer: ByteBuffer): Any? {
        buffer.order(ByteOrder.LITTLE_ENDIAN)
        return read(buffer)
    }

    private fun sizeOf(value: Any?): Int {
        return when (value) {
            null -> 1
            is Boolean -> 2
            is Double -> 9
            is Int -> 9 // We encode all numbers as Double (Float64) for TLV consistency
            is Float -> 9
            is String -> 5 + value.toByteArray(StandardCharsets.UTF_8).size
            is ByteBuffer -> 5 + value.remaining()
            is ByteArray -> 5 + value.size
            is List<*> -> {
                var size = 5 // Type + count
                for (item in value) size += sizeOf(item)
                size
            }
            is Map<*, *> -> {
                var size = 5 // Type + count
                for ((key, item) in value) {
                    size += sizeOf(key) + sizeOf(item)
                }
                size
            }
            else -> 1 // Fallback to null
        }
    }

    private fun write(value: Any?, buffer: ByteBuffer) {
        when (value) {
            null -> buffer.put(0x00.toByte())
            is Boolean -> {
                buffer.put(0x01.toByte())
                buffer.put((if (value) 1 else 0).toByte())
            }
            is Double -> {
                buffer.put(0x03.toByte())
                buffer.putDouble(value)
            }
            is Int -> {
                buffer.put(0x03.toByte())
                buffer.putDouble(value.toDouble())
            }
            is Float -> {
                buffer.put(0x03.toByte())
                buffer.putDouble(value.toDouble())
            }
            is String -> {
                buffer.put(0x04.toByte())
                val bytes = value.toByteArray(StandardCharsets.UTF_8)
                buffer.putInt(bytes.size)
                buffer.put(bytes)
            }
            is ByteBuffer -> {
                buffer.put(0x07.toByte())
                buffer.putInt(value.remaining())
                buffer.put(value)
            }
            is ByteArray -> {
                buffer.put(0x07.toByte())
                buffer.putInt(value.size)
                buffer.put(value)
            }
            is List<*> -> {
                buffer.put(0x05.toByte())
                buffer.putInt(value.size)
                for (item in value) write(item, buffer)
            }
            is Map<*, *> -> {
                buffer.put(0x06.toByte())
                buffer.putInt(value.size)
                // Sort keys for consistency if they are strings
                val keys = value.keys.filterIsInstance<String>().sorted()
                for (key in keys) {
                    write(key, buffer)
                    write(value[key], buffer)
                }
            }
        }
    }

    private fun read(buffer: ByteBuffer): Any? {
        val type = buffer.get().toInt()
        return when (type) {
            0x00 -> null
            0x01 -> buffer.get().toInt() == 1
            0x03 -> buffer.getDouble()
            0x04 -> {
                val len = buffer.getInt()
                val bytes = ByteArray(len)
                buffer.get(bytes)
                String(bytes, StandardCharsets.UTF_8)
            }
            0x07 -> {
                val len = buffer.getInt()
                val bytes = ByteArray(len)
                buffer.get(bytes)
                ByteBuffer.allocateDirect(len).put(bytes).flip()
            }
            0x05 -> {
                val count = buffer.getInt()
                val list = mutableListOf<Any?>()
                for (i in 0 until count) list.add(read(buffer))
                list
            }
            0x06 -> {
                val count = buffer.getInt()
                val map = mutableMapOf<String, Any?>()
                for (i in 0 until count) {
                    val key = read(buffer) as String
                    val value = read(buffer)
                    map[key] = value
                }
                map
            }
            else -> throw Exception("YolkBin: Unknown type $type")
        }
    }
}

File: packages/swift/Sources/Yolk/YolkBin.swift
import Foundation

/// A custom binary protocol (TLV) that matches the TypeScript `YolkBin`.
/// This is the exclusive way data is transferred in Yolk's buffer-only architecture.
public struct YolkBin {
    public static func encode(_ value: Any?) throws -> Data {
        var data = Data()
        try write(value, to: &data)
        return data
    }
    
    public static func decode<T: Decodable>(_ data: Data, as type: T.Type) throws -> T {
        let anyValue = try decode(data)
        if let val = anyValue as? T { return val }
        
        let jsonData = try JSONSerialization.data(withJSONObject: anyValue as Any, options: .fragmentsAllowed)
        return try JSONDecoder().decode(T.self, from: jsonData)
    }
    
    public static func decode(_ data: Data) throws -> Any? {
        var offset = 0
        return try read(from: data, offset: &offset)
    }
    
    private static func write(_ value: Any?, to data: inout Data) throws {
        if value == nil || value is NSNull {
            data.append(0x00)
            return
        }
        
        if let v = value as? Bool {
            data.append(0x01)
            data.append(v ? 1 : 0)
            return
        }
        
        if let v = value as? Double {
            data.append(0x03)
            var val = v.bitPattern.littleEndian
            withUnsafeBytes(of: &val) { data.append(contentsOf: $0) }
            return
        }
        
        if let v = value as? Int {
            data.append(0x03)
            var val = Double(v).bitPattern.littleEndian
            withUnsafeBytes(of: &val) { data.append(contentsOf: $0) }
            return
        }
        
        if let v = value as? String {
            data.append(0x04)
            let utf8 = v.data(using: .utf8)!
            var len = UInt32(utf8.count).littleEndian
            withUnsafeBytes(of: &len) { data.append(contentsOf: $0) }
            data.append(utf8)
            return
        }

        if let v = value as? Data {
            data.append(0x07)
            var len = UInt32(v.count).littleEndian
            withUnsafeBytes(of: &len) { data.append(contentsOf: $0) }
            data.append(v)
            return
        }
        
        if let v = value as? [Any?] {
            data.append(0x05)
            var count = UInt32(v.count).littleEndian
            withUnsafeBytes(of: &count) { data.append(contentsOf: $0) }
            for item in v {
                try write(item, to: &data)
            }
            return
        }
        
        if let v = value as? [String: Any?] {
            data.append(0x06)
            let keys = v.keys.sorted()
            var count = UInt32(keys.count).littleEndian
            withUnsafeBytes(of: &count) { data.append(contentsOf: $0) }
            for key in keys {
                try write(key, to: &data)
                try write(v[key]!, to: &data)
            }
            return
        }

        // FAST PATH: Use custom binary protocol if available
        if let yolkBinEncodable = value as? YolkBinEncodable {
            try write(yolkBinEncodable.yolkBinFields(), to: &data)
            return
        }

        // LAST RESORT: Handle Complex Encodables via JSON bridge (Warning: Performance overhead)
        if let encodable = value as? Encodable {
            print("[YolkBin] Performance Warning: \(type(of: value)) is using JSON detour. Implement YolkBinEncodable for better performance.")
            let jsonData = try JSONEncoder().encode(encodable)
            var dict = try JSONSerialization.jsonObject(with: jsonData, options: .fragmentsAllowed)
            dict = cleanForJSON(dict)
            try write(dict, to: &data)
            return
        }

        throw NSError(domain: "YolkBin", code: 1, userInfo: [NSLocalizedDescriptionKey: "Unsupported type: \(type(of: value))"])
    }
    
    private static func read(from data: Data, offset: inout Int) throws -> Any? {
        guard offset < data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
        let type = data[offset]
        offset += 1
        
        switch type {
        case 0x00: return nil
        case 0x01:
            guard offset < data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let v = data[offset] == 1
            offset += 1
            return v
        case 0x03:
            guard offset + 8 <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let v = data.subdata(in: offset..<offset+8).withUnsafeBytes { $0.load(as: UInt64.self) }.littleEndian
            offset += 8
            return Double(bitPattern: v)
        case 0x04:
            guard offset + 4 <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let len = Int(data.subdata(in: offset..<offset+4).withUnsafeBytes { $0.load(as: UInt32.self) }.littleEndian)
            offset += 4
            guard offset + len <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let str = String(data: data.subdata(in: offset..<offset+len), encoding: .utf8)!
            offset += len
            return str
        case 0x07:
            guard offset + 4 <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let len = Int(data.subdata(in: offset..<offset+4).withUnsafeBytes { $0.load(as: UInt32.self) }.littleEndian)
            offset += 4
            guard offset + len <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let buf = data.subdata(in: offset..<offset+len)
            offset += len
            return buf
        case 0x05:
            guard offset + 4 <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let count = Int(data.subdata(in: offset..<offset+4).withUnsafeBytes { $0.load(as: UInt32.self) }.littleEndian)
            offset += 4
            var arr: [Any?] = []
            for _ in 0..<count {
                arr.append(try read(from: data, offset: &offset))
            }
            return arr
        case 0x06:
            guard offset + 4 <= data.count else { throw NSError(domain: "YolkBin", code: 2, userInfo: [NSLocalizedDescriptionKey: "EOF"]) }
            let count = Int(data.subdata(in: offset..<offset+4).withUnsafeBytes { $0.load(as: UInt32.self) }.littleEndian)
            offset += 4
            var obj: [String: Any?] = [:]
            for _ in 0..<count {
                guard let key = try read(from: data, offset: &offset) as? String else { 
                    throw NSError(domain: "YolkBin", code: 3, userInfo: [NSLocalizedDescriptionKey: "Key is not string"]) 
                }
                obj[key] = try read(from: data, offset: &offset)
            }
            return obj
        default:
            throw NSError(domain: "YolkBin", code: 4, userInfo: [NSLocalizedDescriptionKey: "Unknown type \(type) at offset \(offset-1)"])
        }
    }

    private static func cleanForJSON(_ value: Any) -> Any {
        if let d = value as? Double {
            return (d.isNaN || d.isInfinite) ? NSNull() : d
        }
        if let arr = value as? [Any] {
            return arr.map(cleanForJSON)
        }
        if let dict = value as? [String: Any] {
            return dict.mapValues(cleanForJSON)
        }
        return value
    }
}
